### ISSUE TITLE
Spherical mesh tracklength tally is incorrect

### ISSUE THREAD

**Issue by paulromano** [openmc-dev/openmc#2065]
I was using the new `SphericalMesh` class recently to get a radial flux distribution over a sphere and noticed some weird behavior. After digging into it a little more, I'm convinced that using `SphericalMesh` for tracklength tallies gives incorrect answers. To test my hypothesis, I created a model with a single sphere of r=10 cm and then tallied a spherical mesh with 100 equal-volume shells from r=0 cm to r=5 cm (note that this is *not* the full radius of the model). I did two separate tallies, one with a track-length estimator and one with a collision estimator. The results do not agree. To go one step further, I created a second model where the original cell has been divided up into shells that match the spherical mesh and then tally using a cell filter. The flux results from that match the collision estimator spherical mesh tally. So, that indicates that the tracklength tally using `SphericalMesh` is incorrect. Here are the results:
![flux](https://user-images.githubusercontent.com/743095/167913292-3dd36ce0-9f7f-4bc5-a9cf-aaf49d820d2b.png)
and the corresponding script I used to genreate it:
```Python
import openmc
import numpy as np
import matplotlib.pyplot as plt

mat = openmc.Material()
mat.add_nuclide('U235', 1.0)
mat.set_density('g/cm3', 10.0)

sphere = openmc.Sphere(r=10.0, boundary_type='vacuum')
cell = openmc.Cell(fill=mat, region=-sphere)
model = openmc.Model()
model.geometry = openmc.Geometry([cell])
model.settings.particles = 100_000
model.settings.inactive = 10
model.settings.batches = 20

sph_mesh = openmc.SphericalMesh()
sph_mesh.r_grid = np.linspace(0.0, 5.0**3, 20)**(1/3)
tally1 = openmc.Tally()
tally1.filters = [openmc.MeshFilter(sph_mesh)]
tally1.scores = ['flux']
tally1.estimator = 'collision'

sph_mesh = openmc.SphericalMesh()
sph_mesh.r_grid = np.linspace(0.0, 5.0**3, 20)**(1/3)
tally2 = openmc.Tally()
tally2.filters = [openmc.MeshFilter(sph_mesh)]
tally2.scores = ['flux']
tally2.estimator = 'tracklength'

model.tallies = openmc.Tallies([tally1, tally2])

# Run OpenMC
sp_filename = model.run()

# Get radial flux distribution
with openmc.StatePoint(sp_filename) as sp:
    flux_collision = sp.tallies[tally1.id].mean.ravel()
    flux_tracklength = sp.tallies[tally2.id].mean.ravel()

# Now create model with explicit shells
surfaces = [openmc.Sphere(r=r) for r in sph_mesh.r_grid[1:]] + [sphere]
regions = openmc.model.subdivide(surfaces)[:-1]
cells = [openmc.Cell(fill=mat, region=reg) for reg in regions]
shell_model = openmc.Model()
shell_model.geometry = openmc.Geometry(cells)
shell_model.settings.particles = 100_000
shell_model.settings.inactive = 10
shell_model.settings.batches = 20

# This time use a cell filter
celltally = openmc.Tally()
celltally.filters = [openmc.CellFilter(cells[:-1])]
celltally.scores = ['flux']
shell_model.tallies.append(celltally)

# Run OpenMC and get result
shell_results = shell_model.run()
with openmc.StatePoint(shell_results) as sp:
    flux_shell = sp.tallies[celltally.id].mean.ravel()

# Plot flux distribution
fig, ax = plt.subplots()
ax.stairs(flux_collision, sph_mesh.r_grid, label='Collision')
ax.stairs(flux_tracklength, sph_mesh.r_grid, label='Tracklength')
ax.stairs(flux_shell, sph_mesh.r_grid, label='Cell tally')
ax.legend()
ax.set_xlabel('Radius [cm]')
ax.set_ylabel('Flux')
plt.savefig('flux.png')
```

@ojschumann I haven't dug into the methods on the `SphericalMesh` class much yet to figure out where the bug is. Let me know if you have time to look into this. Also, I haven't tried a similar thing with the `CylindricalMesh` class but I think we ought to check that one too.

**Comment by pshriwise** [comment]
If it's relatively easy, would you mind posting an update of the above plot with the fix in #2108 for closure here @paulromano?

**Comment by paulromano** [comment]
For posterity, here is the same plot as above with the fix from #2108:
![flux](https://user-images.githubusercontent.com/743095/178313959-13c81fab-ec43-4a17-af23-e87075ec58ab.png)
